### PR TITLE
Implement raqote backend for canvas 2D rendering

### LINKED ISSUE #23431: Use raqote for 2D canvas rendering
While working on #22957, @jdm mentioned the possiblity to use [https://github.com/jrmuizel/raqote](https://github.com/jrmuizel/raqote) as a canvas rendering backend.
As stated in [https://github.com/servo/servo/issues/22957#issuecomment-493991851](https://github.com/servo/servo/issues/22957#issuecomment-493991851), this issue is for tracking progress of the conversion to raqote.

Remaining missing pieces:
* [x] LinearGradient/RadialGradient (https://docs.rs/raqote/0.6.4/raqote/struct.Gradient.html)
* [x] ColorBurn support (https://github.com/jrmuizel/raqote/issues/81)
* [x] Path::transformed_copy_to_buiilder (https://github.com/jrmuizel/raqote/issues/82)
* [x] Shadows (https://github.com/jrmuizel/raqote/issues/83)
* [x] `snapshot` API (might just be `SourceSurface::Raqote(self.snapshot_data_owned())`
* [x] `ellipse` path API

### PR DESCRIPTION
**Pull request by pylbrecht**
This is a follow-up of #23936.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #23431

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/24048)

### CODE REVIEW

File: components/canvas/raqote_backend.rs [hunk 1]
```diff
@@ -183,8 +184,28 @@ impl Path {
         unimplemented!()
     }
 
-    pub fn contains_point(&self, _x: f64, _y: f64, _path_transform: &Transform2D<f32>) -> bool {
-        unimplemented!()
+    pub fn contains_point(&self, x: f64, y: f64, _path_transform: &Transform2D<f32>) -> bool {
+        for op in self.as_raqote().ops.iter() {
```
Comment (CYBAI) [inline components/canvas/raqote_backend.rs]: I'm just curious is it possible to run this with [`any`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any) with iterator 👀?

```rust
self.as_raqote().ops.iter().any(|op| match op {
  ...
})
```

Comment (pylbrecht) [reply components/canvas/raqote_backend.rs]: I did not know about `any`. It's exactly what I intended to do, so thanks for pointing it out!

**Review by jrmuizel — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/servo/servo/24048)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @asajeffrey, @CYBAI, and @pylbrecht)

___
*[components/canvas/raqote_backend.rs, line 190 at r3](https://reviewable.io/reviews/servo/servo/24048#-LnD0Mow0ZNUqvpl8WRN:-LnD0Mow0ZNUqvpl8WRO:b6ubx8p) ([raw file](https://github.com/servo/servo/blob/89b8bd516fb2173ecd0f7eef98dc3744a0ddccc3/components/canvas/raqote_backend.rs#L190)):*
> ```Rust
>         self.as_raqote().ops.iter().any(|op| match op {
>             PathOp::MoveTo(point) | PathOp::LineTo(point) => {
>                 point.x as f64 == x && point.y as f64 == y
> ```

This contains_point() implementation is wrong. contains_point() is supposed to return whether the point is inside the path and not whether any of the path ops use that particular point.

**Review by jrmuizel — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/servo/servo/24048)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @asajeffrey, @CYBAI, and @pylbrecht)

___
*[components/canvas/raqote_backend.rs, line 190 at r3](https://reviewable.io/reviews/servo/servo/24048#-LnD0Mow0ZNUqvpl8WRN:-LnD17NFDfb4v4AE0AtB:bpc15rc) ([raw file](https://github.com/servo/servo/blob/89b8bd516fb2173ecd0f7eef98dc3744a0ddccc3/components/canvas/raqote_backend.rs#L190)):*
<details><summary><i>Previously, jrmuizel (Jeff Muizelaar) wrote…</i></summary><blockquote>

This contains_point() implementation is wrong. contains_point() is supposed to return whether the point is inside the path and not whether any of the path ops use that particular point.
</blockquote></details>

I'll add an implementation of contains_point() to raqote.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by pylbrecht** [comment]
> This contains_point() implementation is wrong. contains_point() is supposed to return whether the point is inside the path and not whether any of the path ops use that particular point.

Thanks for pointing that out, @jrmuizel! I have zero experience in implementing 2D graphics whatsoever. I hope this doesn't get in the way too much, while working on this issue. I really don't want to create more work than neccessary.
May I hit you up, if I have any questions regarding 2D graphics?

File: components/canvas/raqote_backend.rs [hunk 2]
```diff
@@ -183,8 +184,9 @@ impl Path {
         unimplemented!()
     }
 
-    pub fn contains_point(&self, _x: f64, _y: f64, _path_transform: &Transform2D<f32>) -> bool {
-        unimplemented!()
+    pub fn contains_point(&self, x: f64, y: f64, _path_transform: &Transform2D<f32>) -> bool {
+        let path = self.as_raqote();
+        path.contains_point(0.1, path.winding, x as f32, y as f32)
```
Comment (jrmuizel) [inline components/canvas/raqote_backend.rs]: The newest version of raqote drops the winding parameter from contains_point() (though you can just deal with that later)

**Review by jrmuizel — APPROVED** [review]
Looks reasonable to me.

File: components/canvas/raqote_backend.rs [hunk 3]
```diff
@@ -488,7 +496,19 @@ impl GenericPathBuilder for PathBuilder {
...
+        let path = self.finish();
+
+        for op in path.as_raqote().ops.iter().rev() {
+            match op {
+                PathOp::MoveTo(point) | PathOp::LineTo(point) => {
+                    return Point2D::new(point.x, point.y)
+                },
+                PathOp::CubicTo(_, _, point) => return Point2D::new(point.x, point.y),
+                PathOp::QuadTo(_, point) => return Point2D::new(point.x, point.y),
+                _ => {},
```
[the remark this answers is not in the document]
Comment (pylbrecht) [reply components/canvas/raqote_backend.rs]: Fixed in d24568218.

**Pull request merged**